modules/blockbuilder/blockbuilder.go

```diff
@@ -205,80 +217,76 @@ func (b *BlockBuilder) consume(ctx context.Context) error {
 	// Clear all previous remnants
 	err := b.wal.Clear()
 	if err != nil {
-		return err
+		return 0, err
 	}
 
-	for _, partition := range partitions {
-		// Consume partition while data remains.
-		// TODO - round-robin one consumption per partition instead to equalize catch-up time.
-		for {
-			more, err := b.consumePartition(ctx, partition, end)
-			if err != nil {
-				return err
-			}
+	ps, err := b.fetchPartitions(ctx, partitions)
+	if err != nil {
+		return 0, err
+	}
 
-			if !more {
-				break
-			}
+	// First iteration over all the assigned partitions to get their current lag in time
+	for i, p := range ps {
+		if !p.hasRecords { // No records, we can skip the partition
+			continue
 		}
+		lastRecordTs, lastRecordOffset, err := b.consumePartition(ctx, p)
+		if err != nil {
+			return 0, err
+		}
+		ps[i].lastRecordTs = lastRecordTs
+		ps[i].startOffset = lastRecordOffset
 	}
 
-	return nil
-}
+	// Iterate over the laggiest partition until the lag is less than the cycle duration or none of the partitions has records
+	for {
+		sort.Slice(ps, func(i, j int) bool {
+			return ps[i].lastRecordTs.After(ps[j].lastRecordTs)
+		})
 
-func getActivePartitions(partitions []int32) string {
-	var strArr []string
-	for _, v := range partitions {
-		strArr = append(strArr, strconv.Itoa(int(v)))
+		laggiestPartition := ps[0]
+		if laggiestPartition.lastRecordTs.IsZero() {
+			return b.cfg.ConsumeCycleDuration, nil
+		}
+
+		lagTime := time.Since(laggiestPartition.lastRecordTs)
+		if lagTime < b.cfg.ConsumeCycleDuration {
+			return time.Second, nil
```

Comment:
Should we make this `lagTime` instead? In my tests when the throughput is high waiting so much time allows the partition to get filled again

Reply:
Yeah, we can try that, it seems like a good solution